src/support/allocators/bulk_pool.h

```diff
@@ -0,0 +1,344 @@
+#include <new>
+#include <type_traits>
+#include <utility>
+
+namespace bulk_pool {
+
+/// This pool allocates one object (memory block) at a time. The first call to allocate determine
+/// the size of chunks in the pool. Subsequent calls with other types of different size will
+/// return nullptr. It is designed for use in node-based container like std::unordered_map or
+/// std::list, where most allocations are for 1 element. It allocates increasingly large blocks
+/// of memory, and only deallocates at destruction.
+///
+/// The size of the allocated blocks are doubled until NUM_CHUNKS_ALLOC_MAX is reached. This way we
+/// don't have too large an overhead for small pool usage, and still get efficiency for lots of
+/// elements since number of mallocs() are very reduced. Also, less bookkeeping is necessary, so
+/// it's more space efficient than allocating one chunk at a time.
+///
+/// Deallocate() does not actually free memory, but puts the data into a linked list which can then
+/// be used for allocate() calls when 1 element is requested. The linked list is in place, reusing
+/// the memory of the chunks.
+///
+/// Memory layout
+///
+///  m_blocks
+///        v
+///    [ BlockNode, chunk0, chunk1, ... chunk3] // block 0
+///        v
+///    [ BlockNode, chunk0, chunk1, ... chunk7] // block 1
+///        :
+///    [ BlockNode, chunk0, chunk1, ... chunk16383] // block 12
+///        v
+///    [ BlockNode, chunk0, chunk1, ... chunk16383] // block 13
+///        v
+///      nullptr
+///
+/// m_free_chunks represents a singly linked list of all T's that have been deallocated.
+class Pool
+{
+public:
+    // Inplace linked list of all allocated blocks. Make sure it is aligned in such a way that whatever comes
+    // after a BlockNode is correctly aligned.
+    struct alignas(alignof(::max_align_t)) BlockNode {
+        // make sure to align
+        BlockNode* next;
+    };
+
+    // Inplace linked list of the allocation chunks
+    struct ChunkNode {
+        ChunkNode* next;
+    };
+
+    /// Explicitly specify the Pool's chunk size. It will only allocate chunks of this size, returning
+    /// nullptr if a type of different size is specified. If set to 0, the first Allocate() call will
+    /// determine the chunk size of the pool.
+    explicit Pool(size_t chunk_size) noexcept
+        : m_chunk_size(chunk_size)
+    {
+    }
+
+    /// Doesn't specify the chunk size, so it's determined by the first call to Allocate.
+    Pool() = default;
+
+    // Don't allow moving/copying a pool, it's dangerous
+    Pool(Pool&&) = delete;
+    Pool& operator=(Pool&&) = delete;
+    Pool(const Pool&) = delete;
+    Pool& operator=(const Pool&) = delete;
+
+    /// Deallocates all allocated memory, even when Deallocate() was not yet called.
+    ~Pool() noexcept
+    {
+        Destroy();
+    }
+
+    /// Don't allow allocation for types that are smaller than a Node. This does not make sense
+    /// because we need to fit a pointer into the memory.
```

Comment:
Would it be better to drop these methods or to use a static assert, so it could be a compile time error when `sizeof(T) < sizeof(ChunkNode)` instead of a runtime error?

Reply:
Thanks for having a look at the code! In that case I think it it better to return `nullptr`, because of how it's used in `bulk_pool::Allocator<T>::allocate(size_t n)`. There `bulk_pool::Pool::Allocate()` is called, and if `nullptr` is returned I fall back to `::operator new`

I could probably rewrite it to be a compile error, but then I need to rewrite `bulk_pool::Allocator<T>::allocate(size_t n)` to handle that case - which I don't know how to do nicely

Reply:
Oh, I see. I didn't realize it would fall back to default operator new. Could mention that explicitly, but this is also fine.